packages/teleport/src/components/StepSlider/StepSlider.tsx

```diff
@@ -0,0 +1,258 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React, { useState, useRef, useEffect, RefObject } from 'react';
+import { TransitionGroup, CSSTransition } from 'react-transition-group';
+import styled from 'styled-components';
+
+export default function StepSlider<T>(props: Props<T>) {
+  const {
+    flows,
+    currFlow,
+    onSwitchFlow,
+    render,
+    tDuration = 500,
+    ...rest
+  } = props;
+
+  // step defines the current step we are in the current flow.
+  const [step, setStep] = useState(0);
+  const [height, setHeight] = useState(0);
+  const [className, setClassName] = useState<'next' | 'prev' | ''>('');
+
+  // preMount is used to invisibly render the next view so we
+  // can get its height. This height is needed in advance
+  // so the height can be animated along with the transform transitions.
+  const [preMount, setPreMount] = useState(false);
+
+  // rootRef is used to set the height on initial render.
+  // Needed to animate the height on initial transition.
+  const rootRef = useRef<HTMLElement>();
+
+  // preMountState is used to hold the latest pre mount data.
+  // useState's could not be used b/c they became stale for
+  // our func 'setHeightOnPreMount'.
+  const preMountState = useRef<{ step: number; flow: keyof T }>({} as any);
+
+  // Sets the initial height.
+  useEffect(() => {
+    if (!rootRef?.current) return;
+    const { height } = rootRef.current.getBoundingClientRect();
+    setHeight(height);
+  }, []);
+
+  // After pre mount, we can calculate the exact height of the next step.
+  // After calculating height, we increment the step to trigger the
+  // animations.
+  const setHeightOnPreMount = (node: HTMLElement) => {
+    if (node !== null) {
+      setHeight(node.getBoundingClientRect().height);
+      setStep(preMountState.current.step);
+      setPreMount(false);
+
+      if (preMountState.current.flow) {
+        onSwitchFlow(preMountState.current.flow);
+      }
+    }
+  };
+
+  function getStep(View: StepComponent, requirePreMount = false) {
+    return (
+      <View
+        key={step}
+        refCallback={requirePreMount ? setHeightOnPreMount : null}
+        onNext={() => {
+          preMountState.current.step = step + 1;
+          setPreMount(true);
+          setClassName('next');
+        }}
+        onPrev={() => {
+          preMountState.current.step = step - 1;
+          setPreMount(true);
+          setClassName('prev');
+        }}
+        onSwitchFlow={(flow, isBack = false) => {
+          preMountState.current.step = 0;
+          preMountState.current.flow = flow;
+
+          setPreMount(true);
+          if (isBack) {
+            setClassName('next');
+            return;
+          }
+          setClassName('prev');
+        }}
+        {...rest}
+      />
+    );
+  }
+
+  let $content;
+  const Step = flows[currFlow][step];
```

Comment:
In order to support the flow being switched from the outside by updating the `currFlow` prop I THINK we need a `useEffect` for `currFlow` so that we can appropriately animate it back to 0?

Reply:
hm, I wasn't sure what you meant. Are you saying currently, switching flow from outside would not work?

Reply:
Well no, it'll work, but I think if you're on flow 1 step 3, then externally change it to flow 2, it'll have to go from flow 1 step 3 to flow 2 step 0. And I don't think this will correctly apply the "prev" animation.

Reply:
oh i think i get what you mean, like this:

flow1: 1, 2, 3, 4a
flow2:             4b

So on step 3, we can go `4a` or `4b`, and be able to go back to step 3?